plans/phase-1/api-design/multigres-operator-api-v1alpha1-design.md

```diff
@@ -0,0 +1,1121 @@
+* **Perpetual GitOps Drift:** Since users can create databases via DDL at any time, the `MultigresCluster` CR in Git will rarely match the actual cluster state. `kubectl diff` will be noisy, and the `status` field will become the only reliable view of the system, degrading the value of the declarative spec. This can be mitigated if we have a component that constantly writes these changes to either git and applies them declaratively, but it is not a common pattern.
+* **Status Object Bloat (Scalability):** Because the Operator must track "Discovered" (DDL-created) databases in the `status` field to make them visible to SREs, a cluster with thousands of databases risks hitting the Kubernetes object size limit (etcd limits). This limits the scalability of the reporting mechanism compared to fanning out to separate CRs. We don't foresee an issue with this at this stage as we are planning to support only one database for the initial MVP, even though the spec has been defined to accept more for future compatibility.
+* **Resource "Adoption" Friction:** Databases created via DDL are assigned a default `ShardTemplate`. To "upgrade" or resize these databases, an SRE must manually "adopt" them by adding them to the `MultigresCluster` YAML with the correct name and new template. This introduces a manual step and a potential race condition where a new database might be under-provisioned before it can be adopted.
+* **Default DB Schema (System Catalog) Availability Dependency:** The Operator's reconciliation loop now strictly depends on the read availability of the "Default Database" (System Catalog). If the Default Database is down or locked, the Operator becomes "blind" and cannot reconcile *any* part of the cluster, unlike a pure CRD model which relies only on the highly available Kubernetes API. This may not be much of an issue as either way, if the default DB is down, there it wouldn't work either way. 
+* **API Hotspot and "Blast Radius" Risk:** By centralizing all database definitions into the monolithic `MultigresCluster` CR, this single object becomes a massive reconciliation hotspot. A single typo in this large resource (e.g., while "adopting" a database) could potentially break reconciliation for the entire cluster's control plane.
+* **Rename/Replace Ambiguity:** Postgres allows `ALTER DATABASE RENAME`, but Kubernetes relies on stable names. If a user renames a database in SQL, the Operator may perceive this as a "Delete" (of the old name) and "Create" (of the new name), potentially attempting to re-provision the old name if it still exists in the YAML.
+* **Lack of Namespace Isolation:** All databases must be defined in (or adopted into) the central `MultigresCluster` resource. This effectively forces all application teams to rely on platform admins to manage resource sizing, removing the ability to use Kubernetes RBAC for self-service resource management in separate namespaces. No clean DBA/Platform Engineer persona separation model. 
```

Comment:
Could you explain this more in detail? The namespace will be tied to the target MultigresCluster, but that's somewhat expected I thought. That's the normal multi-tenancy deployment, where each namespace has a list of resources.

Reply:
In the alternative DDL design you could have MultigresCluster in one namespace and the DBs on other namespaces, effectively separated DBAs/App devs from SRE/Platform Engineers. I suppose it's not really a drawback if the use case makes sense to design it without this separation, which it does, at least for now.